File: clients/python/highbar_client/state_stream.py
# SPDX-License-Identifier: GPL-2.0-only
"""StreamState consumer with seq-monotonicity checker (T087).

Port of clients/fsharp/src/StateStream.fs. The invariant checker is
authoritative: any gap, duplicate, or regression raises
``SeqInvariantError`` — equivalent to the F# client's
``SeqInvariantException``, so SC-004's byte-equality claim between
F# and Python clients extends to their error paths too.
"""

from __future__ import annotations

from collections.abc import Iterator
from typing import Optional

import grpc

from .highbar import service_pb2, service_pb2_grpc, state_pb2  # type: ignore


class SeqInvariantError(RuntimeError):
    """Raised when a StateUpdate has non-monotonic seq."""


def consume(
    channel: grpc.Channel,
    resume_from_seq: int = 0,
    max_wait_seconds: Optional[float] = None,
) -> Iterator[state_pb2.StateUpdate]:
    """Yield StateUpdates from StreamState. Checks monotonic seq.

    The caller iterates and runs their own per-update logic. Cancel
    by closing the channel or breaking out of the loop — grpcio
    sends CANCELLED to the server on iterator close.
    """
    stub = service_pb2_grpc.HighBarProxyStub(channel)
    req = service_pb2.StreamStateRequest(resume_from_seq=resume_from_seq)

    last_seq: Optional[int] = None
    call_kwargs: dict = {}
    if max_wait_seconds is not None:
        call_kwargs["timeout"] = max_wait_seconds

    for update in stub.StreamState(req, **call_kwargs):
        if last_seq is not None and update.seq <= last_seq:
            raise SeqInvariantError(
                f"seq regression: got {update.seq} after {last_seq}"
            )
        last_seq = update.seq
        yield update


def record(
    channel: grpc.Channel,
    resume_from_seq: int = 0,
    max_updates: Optional[int] = None,
    max_wait_seconds: Optional[float] = None,
) -> list[state_pb2.StateUpdate]:
    """Diagnostic: drain up to ``max_updates`` updates into a list.

    Do not use on long-lived streams — the buffer is unbounded except
    for ``max_updates``.
    """
    out: list[state_pb2.StateUpdate] = []
    try:
        for update in consume(channel, resume_from_seq, max_wait_seconds):
            out.append(update)
            if max_updates is not None and len(out) >= max_updates:
                break
    except grpc.RpcError as e:
        # Deadline hits here; callers usually want partial results
        # rather than losing them.
        if e.code() != grpc.StatusCode.DEADLINE_EXCEEDED:
            raise
    return out

File: clients/fsharp/src/StateStream.fs
// SPDX-License-Identifier: GPL-2.0-only
//
// HighBarV3 — StreamState consumer (T043, T046).
//
// Wraps the server-streaming StreamState RPC with a seq-invariant
// checker (FR-006: strictly monotonic; SC-005: no gaps, no dupes).
// Emits each StateUpdate to the caller's handler as it arrives.
//
// The invariant checker is authoritative — if the server-side
// ring/replay ever drifts it will fail here loudly rather than
// downstream in observer UIs.

namespace HighBar.Client

open System
open System.Collections.Generic
open System.Threading
open System.Threading.Tasks
open Grpc.Core
open Grpc.Net.Client
open Highbar.V1

module StateStream =

    /// Callback invoked once per StateUpdate. Runs on the gRPC response
    /// thread — handlers should be cheap or marshal onto their own
    /// dispatcher.
    type OnUpdate = StateUpdate -> unit

    /// Seq-invariant violation details — thrown when the client
    /// detects a gap, duplicate, or out-of-order seq.
    exception SeqInvariantException of message: string

    /// Open a StreamState and pump updates into `onUpdate` until the
    /// stream closes (normal) or the cancellation token fires.
    /// `resumeFromSeq = 0u` asks for a fresh snapshot; any other
    /// value is a resume request (data-model §2).
    let consume (channel: GrpcChannel)
                (resumeFromSeq: uint64)
                (onUpdate: OnUpdate)
                (ct: CancellationToken)
                : Async<unit> =
        async {
            let client = HighBarProxy.HighBarProxyClient(channel)
            let req = StreamStateRequest(ResumeFromSeq = resumeFromSeq)
            use call = client.StreamState(req, cancellationToken = ct)

            let mutable lastSeq : uint64 voption = ValueNone
            let mutable snapshotsSeen = 0

            let! _ =
                async {
                    while! call.ResponseStream.MoveNext(ct) |> Async.AwaitTask do
                        let upd = call.ResponseStream.Current

                        match lastSeq with
                        | ValueSome prev when upd.Seq <= prev ->
                            raise (SeqInvariantException
                                    (sprintf
                                        "seq regression: got %d after %d"
                                        upd.Seq prev))
                        | _ -> ()

                        // A snapshot arm that is NOT the first message
                        // in the stream indicates a server-side reset
                        // (resume out of range → fresh snapshot). That
                        // is legal and monotonic-continuous by
                        // construction on the server (data-model §2
                        // invariant: snapshot reset preserves seq
                        // monotonicity); we just count them so tests
                        // can assert.
                        if upd.PayloadCase = StateUpdate.PayloadOneofCase.Snapshot then
                            snapshotsSeen <- snapshotsSeen + 1

                        lastSeq <- ValueSome upd.Seq
                        onUpdate upd
                }
            return ()
        }

    /// Diagnostic wrapper — runs `consume` and records the stream as a
    /// list for assertions. Do not use for long streams (unbounded
    /// memory).
    let record (channel: GrpcChannel)
               (resumeFromSeq: uint64)
               (ct: CancellationToken)
               : Async<ResizeArray<StateUpdate>> =
        async {
            let buf = ResizeArray<StateUpdate>()
            do! consume channel resumeFromSeq buf.Add ct
            return buf
        }
